File: Python/programmers/etc/level2/호텔_대실.py
from heapq import *

def calc(time):
    h, m = map(int, time.split(":"))
    return h*60+m

def solution(book_time):
    q = []
    for s, e in book_time:
        s, e = calc(s), calc(e)
        heappush(q, [s, e])
    
    last = [-10]
    while q:
        cur_s, cur_e = heappop(q)
        if cur_s >= last[0]+10:
            heappop(last)
            heappush(last, cur_e)
        else:
            heappush(last, cur_e)
            
    return len(last)


book_time = [["15:00", "17:00"], ["16:40", "18:20"], ["14:20", "15:20"], ["14:10", "19:20"], ["18:20", "21:20"]]
print(solution(book_time)) # 3

book_time = [["09:10", "10:10"], ["10:20", "12:20"]]
print(solution(book_time)) # 1

book_time = [["10:20", "12:30"], ["10:20", "12:30"], ["10:20", "12:30"]]
print(solution(book_time)) # 3

File: Java/programmers/etc/level2/호텔대실.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 호텔대실 {
    public int solution(String[][] book_time) {
        PriorityQueue<int[]> q = new PriorityQueue<>((o1, o2) -> {return o1[0]-o2[0];});
        for (String[] book : book_time) {
            String s = book[0], e = book[1];
            q.add(new int[]{calc(s), calc(e)});
        }
        PriorityQueue<Integer> last = new PriorityQueue<>();
        last.add(10);
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cur_s = cur[0], cur_e = cur[1];
            if (cur_s >= last.peek()+10) {
                last.poll();
            }
            last.add(cur_e);
        }
        return last.size();
    }
    
    private int calc(String str) {
        String[] st = str.split(":");
        return Integer.parseInt(st[0])*60 + Integer.parseInt(st[1]);
    }
}
